### ISSUE TITLE
Consensus issue with how/ what CALL operations touch accounts to be deleted

### ISSUE THREAD

**Issue by soc1c** [etclabscore/go-ethereum#25]
For context, EIP 158/161 touches any accounts through any CALL type operations with a zero value transfer (CALL, DELEGATECALL, CALLCODE), which marks them for deletion at the end of the transaction if they are still empty objects. Empty objects is defined as having 0 balance, nonce, code, and storage.

Currently in ETH Geth, [there is an intentional touch on STATICCALL implementation](https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L344) that causes a touch (there is no functional need for this because STATICCALL is explicitly a non-state changing operation) but also CALLCODE and DELEGATECALL currently on ETC transfer a value of 0 since it is a shared `exec(...)` function, which causes a touch. This did not matter previously (or when all empty object will be deleted) because the behaviour would be consistent. 

The consensus issue made most explicit is if the clients make a STATICCALL operation to an empty account at an address and that transaction runs OOG, the different clients will disagree with which empty accounts will be deleted. This is not obvious by tests because the tests prior and including Byzantium did not expose this change, but a lot of the Constantinople tests do (maybe a reason to follow this strange implementation)

On ETH Geth, they are currently expecting the failure of certain tests because of their implementation:

```go
	// Expected failures:
	st.fails(`^stRevertTest/RevertPrecompiledTouch(_storage)?\.json/Byzantium/0`, "bug in test")
	st.fails(`^stRevertTest/RevertPrecompiledTouch(_storage)?\.json/Byzantium/3`, "bug in test")
	st.fails(`^stRevertTest/RevertPrecompiledTouch(_storage)?\.json/Constantinople/0`, "bug in test")
	st.fails(`^stRevertTest/RevertPrecompiledTouch(_storage)?\.json/Constantinople/3`, "bug in test")
	st.fails(`^stRevertTest/RevertPrecompiledTouch(_storage)?\.json/ConstantinopleFix/0`, "bug in test")
	st.fails(`^stRevertTest/RevertPrecompiledTouch(_storage)?\.json/ConstantinopleFix/3`, "bug in test")
```

For reference here is what these tests call inside the test in assembly:

(Explained in plain english each test is calling all of the precompile addresses with 0 value with different CALL operation types)

0: CALL)
```go
"code" : "{ (CALL 50000 1 0 0 0 0 0) (CALL 50000 2 0 0 0 0 0) (CALL 50000 3 0 0 0 0 0) (CALL 50000 4 0 0 0 0 0) (CALL 50000 5 0 0 0 0 0) (CALL 50000 6 0 0 0 0 0) (CALL 50000 7 0 0 0 0 0) (CALL 50000 8 0 0 0 0 0) [[1]] (GAS) [[2]] (GAS) [[3]] (GAS) }",
```

3: STATICCALL)
```go
"code" : "{ (STATICCALL 50000 1 0 0 0 0) (STATICCALL 50000 2 0 0 0 0) (STATICCALL 50000 3 0 0 0 0) (STATICCALL 50000 4 0 0 0 0) (STATICCALL 50000 5 0 0 0 0) (STATICCALL 50000 6 0 0 0 0) (STATICCALL 50000 7 0 0 0 0) (STATICCALL 50000 8 0 0 0 0) [[1]] (GAS) [[2]] (GAS) [[3]] (GAS) }",
```

and for reference, the two tests that pass and expect to pass:

1: DELEGATECALL)
```go
"code" : "{ (DELEGATECALL 50000 1 0 0 0 0) (DELEGATECALL 50000 2 0 0 0 0) (DELEGATECALL 50000 3 0 0 0 0) (DELEGATECALL 50000 4 0 0 0 0) (DELEGATECALL 50000 5 0 0 0 0) (DELEGATECALL 50000 6 0 0 0 0) (DELEGATECALL 50000 7 0 0 0 0) (DELEGATECALL 50000 8 0 0 0 0) [[1]] (GAS) [[2]] (GAS) [[3]] (GAS) }",
```

2: CALLCODE)
```go
"code" : "{ (CALLCODE 50000 1 0 0 0 0 0) (CALLCODE 50000 2 0 0 0 0 0) (CALLCODE 50000 3 0 0 0 0 0) (CALLCODE 50000 4 0 0 0 0 0) (CALLCODE 50000 5 0 0 0 0 0) (CALLCODE 50000 6 0 0 0 0 0) (CALLCODE 50000 7 0 0 0 0 0) (CALLCODE 50000 8 0 0 0 0 0) [[1]] (GAS) [[2]] (GAS) [[3]] (GAS) }",
```

Which all four of these tests pass if the explicit touch on ETH is removed. 

### To clarify, this implementation caused the tests to fail and ETH expected those tests to fail in order to remain in consensus with Parity on the main net chain

This is a bad reason to follow this implementation, but would make testing Agharta much easier since all tests expect this behaviour now, and it does not functionally change **anything** once all existing empty objects are deleted (though it may have an effect if somehow a way to initialize empty accounts surfaces through a bug).

Another issue is that currently the implementations in Parity and ETH geth is that address 0x00...03 an exception to which addresses are cleared for being empty due to the issue with that precompile being deleted on main net when all empty accounts were touch deleted after the implementation of 161. 
https://github.com/ethereum/go-ethereum/blob/master/core/state/journal.go#L165
This edge case was implemented at some point in some in progress merges into ETC geth despite the fact that it had no meaning on Ethereum Classic network.

[Discussion of related topics referenced, on main net](https://github.com/ethereum/EIPs/issues/716)

Tying into this, we see a similar potential issue in address 0x5, where it [may have been initialized from non-existant -> empty](http://gastracker.io/addr/0x0000000000000000000000000000000000000005) this would present a similar issue to what occured with address 0x3 if it is subject to the same edge case if we would presumably want to align ourselves with the parity implementation? (this could be resolved by sending a minimal balance to this address)

---

The implementation we have now produces the exact same test failures so on second thought there may not be any consensus issues. We inadvertently touched STATICCALL addresses in the exec call because of how the ETC code is structured. We should still talk about refactoring the execution.go file which handles these CALL code logic and be sure we want STATICCALL to touch the addresses it references so there is no consensus issue. I can explain more in detail if this is confusing.

**Comment by soc1c** [comment]
Imported from https://github.com/eth-classic/go-ethereum/issues/41

**Comment by austinabell** [comment]
Behaviour matches other clients and is already addressed. 

I still disagree with the STATICCALL touch as it is defined that: "An account is considered to be touched when it is involved in any potentially state-changing operation" [link](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-161.md) and STATICCALL is by definition a non state changing operation.

Close if you want though because there is no bug here and other clients have this logic included.

**Comment by soc1c** [comment]
ok
